### PR TITLE
Port vectorstrength, coherence and spectrogram from CuSignal to cupyx.scipy.signal

### PR DESCRIPTION
**Pull request by andfoy**
See https://github.com/cupy/cupy/issues/7403
Depends on https://github.com/cupy/cupy/pull/7838

### CODE REVIEW

**Comment by takagi** [comment]
Would you add these APIs to signal.rst? Then, I will kick the CI again to merge this.

File: tests/cupyx_tests/scipy_tests/signal_tests/test_spectral.py [hunk 1]
```diff
@@ -1229,3 +1229,187 @@ def test_roundtrip_scaling(self, xp, scp):
...
+
+    @testing.numpy_cupy_allclose(scipy_name='scp')
+    def test_partial_2dperiod(self, xp, scp):
+        events = xp.array([.25, .5, .75])
+        period = [1., 1., 1., 1.]
+
+        strength, phase = scp.signal.vectorstrength(events, period)
+        return strength, phase
+
+    @testing.numpy_cupy_allclose(scipy_name='scp')
```
Comment (takagi) [inline tests/cupyx_tests/scipy_tests/signal_tests/test_spectral.py:1299]: minor tolerance issue, e.g. https://ci.preferred.jp/cupy.linux.cuda121/144442/
```suggestion
    @testing.numpy_cupy_allclose(scipy_name='scp', rtol=1e-7, atol=1e-7)
```

File: tests/cupyx_tests/scipy_tests/signal_tests/test_spectral.py [hunk 2]
```diff
@@ -1229,3 +1229,187 @@ def test_roundtrip_scaling(self, xp, scp):
...
+    def test_opposite_1dperiod(self, xp, scp):
+        events = xp.array([0, .25, .5, .75])
+        period = 1.
+
+        strength, phase = scp.signal.vectorstrength(events, period)
+        return strength, phase
+
+    @testing.numpy_cupy_allclose(scipy_name='scp')
```
Comment (takagi) [inline tests/cupyx_tests/scipy_tests/signal_tests/test_spectral.py:1307]: ditto (minor tolerance issue, e.g. https://ci.preferred.jp/cupy.linux.cuda121/144442/)
```suggestion
    @testing.numpy_cupy_allclose(scipy_name='scp', rtol=1e-7, atol=1e-7)
```

File: cupyx/scipy/signal/_spectral.py [hunk 3]
```diff
@@ -1136,3 +1136,443 @@ def istft(
...
+           vector strength: Auditory system, electric fish, and noise.
+           Chaos 21, 047508 (2011);
+           :doi:`10.1063/1.3670512`.
+    .. [2] van Hemmen, JL. Vector strength after Goldberg, Brown, and
+           von Mises: biological and mathematical perspectives.  Biol Cybern.
+           2013 Aug;107(4):385-96. :doi:`10.1007/s00422-013-0561-7`.
+    .. [3] van Hemmen, JL and Vollmayr, AN.  Resonating vector strength:
+           what happens when we vary the "probing" frequency while keeping
+           the spike times fixed.  Biol Cybern. 2013 Aug;107(4):491-94.
+           :doi:`10.1007/s00422-013-0560-8`.
```
Comment (takagi) [inline cupyx/scipy/signal/_spectral.py:1548]: Let me remove DOI role to build the doc.
```suggestion
    .. [1] van Hemmen, JL, Longtin, A, and Vollmayr, AN. Testing resonating
           vector strength: Auditory system, electric fish, and noise.
           Chaos 21, 047508 (2011).
    .. [2] van Hemmen, JL. Vector strength after Goldberg, Brown, and
           von Mises: biological and mathematical perspectives.  Biol Cybern.
           2013 Aug;107(4):385-96.
    .. [3] van Hemmen, JL and Vollmayr, AN.  Resonating vector strength:
           what happens when we vary the "probing" frequency while keeping
           the spike times fixed.  Biol Cybern. 2013 Aug;107(4):491-94.
```

**Comment by takagi** [comment]
Would you check the CI result? It still seems to have relatively larger differences even with newer scipy.

```
_tests/scipy_tests/signal_tests/test_spectral.py::TestVectorstrength::test_opposite_1dperiod	
_tests/scipy_tests/signal_tests/test_spectral.py::TestVectorstrength::test_opposite_2dperiod
```

```
AssertionError:	
Not equal to tolerance rtol=1e-07, atol=1e-07	

Mismatched elements: 10 / 10 (100%)	
Max absolute difference: 0.32175055	
Max relative difference: 0.1201483	
 x: array([2.356194, 2.356194, 2.356194, 2.356194, 2.356194, 2.356194,	
       2.356194, 2.356194, 2.356194, 2.356194])	
 y: array([2.677945, 2.677945, 2.677945, 2.677945, 2.677945, 2.677945,	
       2.677945, 2.677945, 2.677945, 2.677945])
```

**Comment by andfoy** [comment]
imagen

I cannot reproduce this locally. I'm using SciPy 1.11.3

File: cupyx/scipy/signal/_spectral.py [hunk 4]
```diff
@@ -1136,3 +1136,441 @@ def istft(
...
+    # the vector phase is the angle of the mean of the vectors
+    vectormean = cupy.mean(vectors, axis=1)
+    strength = cupy.abs(vectormean)
+    phase = cupy.angle(vectormean)
+
+    # if the original period was a scalar, return scalars
+    if scalarperiod:
+        strength = strength[0]
+        phase = phase[0]
+    return strength, phase
```
Comment (andfoy) [inline cupyx/scipy/signal/_spectral.py:1576]: What is strange is that the `vectorstrength` implementation does not contain any custom kernels and it is the same as the SciPy one

Comment (takagi) [reply cupyx/scipy/signal/_spectral.py:1576]: Well, I also can't reproduce this failure on my local.

**Comment by takagi** [comment]
Record: full test result
https://ci.preferred.jp/p/dashboard_by_commit_id?commit_id=a54fcce947623353e474c9bf46e555f2cfc77d28

**Comment by takagi** [comment]
Thanks for the marks on the tests. It seems that the problematic tests pass on linux with `scipy==1.8` (`cupy.linux.cuda113`). Because I'd like to minimize the conditions to mark as xfail, would you change the xfail condition to on linux with `sicpy>=1.9` only? Then, I will kick the full test again and merge this pull request.

**Comment by takagi** [comment]
I investigated the case in the CI environment for a short time, but couldn't reproduce the error.

File: tests/cupyx_tests/scipy_tests/signal_tests/test_spectral.py [hunk 5]
```diff
@@ -1229,3 +1230,193 @@ def test_roundtrip_scaling(self, xp, scp):
...
+    def test_partial_2dperiod(self, xp, scp):
+        events = xp.array([.25, .5, .75])
+        period = [1., 1., 1., 1.]
+
+        strength, phase = scp.signal.vectorstrength(events, period)
+        return strength, phase
+
+    @pytest.mark.xfail(
+        condition=os.name == 'linux' and testing.installed('scipy>=1.9.0'),
```
Comment (takagi) [inline tests/cupyx_tests/scipy_tests/signal_tests/test_spectral.py:1301]: Seems that `os.name` returns `posix` on linux (or at least Ubuntu on my local). Would you use `sys.platform` instead?

File: tests/cupyx_tests/scipy_tests/signal_tests/test_spectral.py [hunk 6]
```diff
@@ -1229,3 +1230,195 @@ def test_roundtrip_scaling(self, xp, scp):
...
+    def test_partial_2dperiod(self, xp, scp):
+        events = xp.array([.25, .5, .75])
+        period = [1., 1., 1., 1.]
+
+        strength, phase = scp.signal.vectorstrength(events, period)
+        return strength, phase
+
+    @pytest.mark.xfail(
+        condition=(
+            sys.platform == 'linux' and testing.installed('scipy>=1.9.0')),
```
Comment (takagi) [inline tests/cupyx_tests/scipy_tests/signal_tests/test_spectral.py:1302]:
```suggestion
        condition=(
            sys.platform == 'linux' and
            11030 <= runtime.runtimeGetVersion() < 11040 and
            testing.installed('scipy==1.8')
        ),
```

File: tests/cupyx_tests/scipy_tests/signal_tests/test_spectral.py [hunk 7]
```diff
@@ -1229,3 +1230,195 @@ def test_roundtrip_scaling(self, xp, scp):
...
+    def test_opposite_1dperiod(self, xp, scp):
+        events = xp.array([0, .25, .5, .75])
+        period = 1.
+
+        strength, phase = scp.signal.vectorstrength(events, period)
+        return strength, phase
+
+    @pytest.mark.xfail(
+        condition=(
+            sys.platform == 'linux' and testing.installed('scipy>=1.9.0')),
```
Comment (takagi) [inline tests/cupyx_tests/scipy_tests/signal_tests/test_spectral.py:1314]:
```suggestion
        condition=(
            sys.platform == 'linux' and
            11030 <= runtime.runtimeGetVersion() < 11040 and
            testing.installed('scipy==1.8')
        ),
```

**Comment by takagi** [comment]
For record: https://ci.preferred.jp/p/dashboard_by_commit_id?commit_id=e6ef62ecd79d28f68eed3bac03995b9be1b06e90

**Review by takagi — COMMENTED** [review]
Let us have this strange condition for the moment to pass CI...

**Comment by takagi** [comment]
Record: https://ci.preferred.jp/p/dashboard_by_commit_id?commit_id=aeb198931458564f4cfade8b1265908580d0c9dd

**Comment by takagi** [comment]
LGTM!

**Review by takagi — APPROVED** [review]

**Comment by takagi** [comment]
Record: https://ci.preferred.jp/p/dashboard_by_commit_id?commit_id=e5ef6ab2ca5678457fc618136dbaf807d5f8683f

**Comment by takagi** [comment]
Thanks, @andfoy!

**Pull request merged**